### ISSUE TITLE
Add support for multithreaded Cell builds

### ISSUE THREAD

**Issue by fatcerberus** [spheredev/neosphere#154]
This would allow Cell to take advantage of multiple cores.

One major challenge here is that multiple threads can't share a Duktape context.  Each thread needs its own JS environment, a requirement which clashes with the current Cell architecture wherein a `Tool` just encapsulates a JS function to be called later by the build engine.

**Comment by rhuanjl** [comment]
Unsure on this one - needs us to be doing a lot of work in the cellscript to see much gain, I guess transpiling ES6->ES5 in duktape with Babel or TypeScript probably wanted this but now I figure the heaviest thing in cell is my .rmp->.mem script and it's pretty quick, I guess multi-threading could tak it from ~10 seconds to convert 50 maps down to 4 or 5 seconds; certainly wouldn't consider it a priority at the moment.

(Maybe if we bring in typescript then typescript + resource conversion would start to want this).

**Comment by fatcerberus** [comment]
Another issue is overhead - because of the way JS  is designed, communication between threads has to involve serialization; we'd effectively be implementing workers.  So we could end up losing a lot of the benefit of parallelism that way...

**Comment by rhuanjl** [comment]
Could you make different threads build different targets so they don't need to talk apart from being told what target to build? (think the runTool function from here https://github.com/rhuanjl/SphereLibs/blob/master/Templates/Cellscript.mjs#L21-L35
But handing out each stage/build instruction to a different thread (or at least queuing them between multiple threads, probably not one per target, maybe sharing them out between 3 or 4).

**Comment by fatcerberus** [comment]
In hindsight, I believe this problem will be intractable for as long as Cell is JavaScript-based.  A `Tool`, by design, encapsulates some JS code which handles the actual build step, at present in the form of a function.  In order to build different targets on different threads, that function would need to be able to be called on separate threads--possibly many instances at once--and JS is not threadsafe.

Long story short, I'm just going to write this off as impractical and close the issue.
